src/engine/qcommon/net_ip.cpp

```diff
@@ -181,150 +195,18 @@ NET_ErrorString
 */
 const char *NET_ErrorString()
 {
-#ifdef _WIN32
+	const auto errorCode = socketError;
 
-	//FIXME: replace with FormatMessage?
-	switch ( socketError )
+	if (errorCode == 0)
 	{
-		case WSAEINTR:
-			return "WSAEINTR";
-
-		case WSAEBADF:
-			return "WSAEBADF";
-
-		case WSAEACCES:
-			return "WSAEACCES";
-
-		case WSAEDISCON:
-			return "WSAEDISCON";
-
-		case WSAEFAULT:
-			return "WSAEFAULT";
-
-		case WSAEINVAL:
-			return "WSAEINVAL";
-
-		case WSAEMFILE:
-			return "WSAEMFILE";
-
-		case WSAEWOULDBLOCK:
-			return "WSAEWOULDBLOCK";
-
-		case WSAEINPROGRESS:
-			return "WSAEINPROGRESS";
-
-		case WSAEALREADY:
-			return "WSAEALREADY";
-
-		case WSAENOTSOCK:
-			return "WSAENOTSOCK";
-
-		case WSAEDESTADDRREQ:
-			return "WSAEDESTADDRREQ";
-
-		case WSAEMSGSIZE:
-			return "WSAEMSGSIZE";
-
-		case WSAEPROTOTYPE:
-			return "WSAEPROTOTYPE";
-
-		case WSAENOPROTOOPT:
-			return "WSAENOPROTOOPT";
-
-		case WSAEPROTONOSUPPORT:
-			return "WSAEPROTONOSUPPORT";
-
-		case WSAESOCKTNOSUPPORT:
-			return "WSAESOCKTNOSUPPORT";
-
-		case WSAEOPNOTSUPP:
-			return "WSAEOPNOTSUPP";
-
-		case WSAEPFNOSUPPORT:
-			return "WSAEPFNOSUPPORT";
-
-		case WSAEAFNOSUPPORT:
-			return "WSAEAFNOSUPPORT";
-
-		case WSAEADDRINUSE:
-			return "WSAEADDRINUSE";
-
-		case WSAEADDRNOTAVAIL:
-			return "WSAEADDRNOTAVAIL";
-
-		case WSAENETDOWN:
-			return "WSAENETDOWN";
-
-		case WSAENETUNREACH:
-			return "WSAENETUNREACH";
-
-		case WSAENETRESET:
-			return "WSAENETRESET";
-
-		case WSAECONNABORTED:
-			return "WSWSAECONNABORTEDAEINTR";
-
-		case WSAECONNRESET:
-			return "WSAECONNRESET";
-
-		case WSAENOBUFS:
-			return "WSAENOBUFS";
-
-		case WSAEISCONN:
-			return "WSAEISCONN";
-
-		case WSAENOTCONN:
-			return "WSAENOTCONN";
-
-		case WSAESHUTDOWN:
-			return "WSAESHUTDOWN";
-
-		case WSAETOOMANYREFS:
-			return "WSAETOOMANYREFS";
-
-		case WSAETIMEDOUT:
-			return "WSAETIMEDOUT";
-
-		case WSAECONNREFUSED:
-			return "WSAECONNREFUSED";
-
-		case WSAELOOP:
-			return "WSAELOOP";
-
-		case WSAENAMETOOLONG:
-			return "WSAENAMETOOLONG";
-
-		case WSAEHOSTDOWN:
-			return "WSAEHOSTDOWN";
-
-		case WSASYSNOTREADY:
-			return "WSASYSNOTREADY";
-
-		case WSAVERNOTSUPPORTED:
-			return "WSAVERNOTSUPPORTED";
-
-		case WSANOTINITIALISED:
-			return "WSANOTINITIALISED";
-
-		case WSAHOST_NOT_FOUND:
-			return "WSAHOST_NOT_FOUND";
-
-		case WSATRY_AGAIN:
-			return "WSATRY_AGAIN";
-
-		case WSANO_RECOVERY:
-			return "WSANO_RECOVERY";
-
-		case WSANO_DATA:
-			return "WSANO_DATA";
-
-		default:
-			return "NO ERROR";
+		return "NO ERROR";
 	}
 
-#else
-	return strerror( errno );
-#endif
+	static thread_local std::string errorMessage;
+	
+	errorMessage = std::system_category().message( errorCode );
```

Comment:
Why would it work to cast the result of `WSAGetLastError()` to a `system_category` code?

Reply:
Based on https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsagetlasterror

"The Windows Sockets extended error codes returned by `WSAGetLastError` and the text description of the error are listed under Windows Sockets Error Codes. These error codes and a short text description associated with an error code are defined in the `Winerror.h` header file. The `FormatMessage` function can be used to obtain the message string for the returned error."

`FormatMessage` in such case can be called only with `FORMAT_MESSAGE_FROM_SYSTEM` which means the function should search the system message-table resource(s) for the requested message. As `std::system_category` is defined as "static error category object for errors reported by the operating system ... It is also required to map the error codes that match POSIX `errno` values to `std::generic_category`." than it is safe to use `std::system_category` for both WIndows Sockets errors and POSIX errors from original Berkeley sockets spec.

Reply:
#include <system_error>
#include <iostream>
#include <Windows.h>


int main() {
  for (int e = WSABASEERR; e <= WSA_IPSEC_NAME_POLICY_ERROR; e++)
    std::cout << e << ' ' << std::system_category().message(e) << '\n';
}

Reply:
![image](https://user-images.githubusercontent.com/1983411/125332920-49ce1c80-e352-11eb-80a1-3fd851bebf4b.png)

Reply:
Note gaps in error codes exactly as defined in `winerror.h`

Reply:
Ah, I should have guessed that it is working with Visual Studio. I tested with mingw, where it does not work. We use mingw for releases, so unfortunately we can't take advantage of these error messages.

Reply:
Hm, strange, it means MinGW STL doesn't follow Visual Studio way.

What if we wrap `system_category` into custom wrapper. and call `system_category` functions when using Visual Studio and have custom implementation via `FormatMessage` for MinGW?

That allows us to simplify switching to a single `system_category` when MinGW follows Visual Studio way.

Reply:
I'd rather not have more implementations than necessary. If the one for mingw can also work for VS then it should be used in both.

Reply:
Smth like this https://github.com/DaemonEngine/Daemon/pull/457/commits/9250a4dbe91651c6628e3e6b0aea73337b98d9c7

Will merge with first error handling commit if all is ok.

Reply:
Oh, that's nice if `FormatMessage` handles `WSAE`* - we already have a `Sys::Win32StrError` function to get the string (removing spaces at the end and so on).

The error category API is unpopular (enough so that [no one noticed it was completely broken by a C++ ABI migration](https://github.com/Unvanquished/Unvanquished/issues/1215#issuecomment-706181341)). So I wouldn't count on anyone adding such functionality. In my view we should try to have only one implementation of a function as much as possible, unless there is a compelling reason to have multiple. One `Win32StrError`-based implementation for all Windows builds would be great.

Reply:
Ok, how about this one: https://github.com/DaemonEngine/Daemon/pull/457/commits/80fa21b3f6e9693c0043a4d52770470cdea28bbb

Unified system error API.

Reply:
Fixed errors, please, take a look.